streaming/time-based moving buffer/dcfbbb48-e17e-44ae-a18d-ee4195cd0b75.json

```diff
@@ -0,0 +1,54 @@
+{
+  "implementations": {
+    "es6": {
+      "initFunction": "state.buffer = [];",
+      "inputFunctions": {
+        "7a2da0d9-b262-4d23-90b3-55a833924d8f": "const {size} = params;\nif (typeof size !== \"number\" || size === 0) {\n  outputs.error({message: \"Invalid buffer size\"}, tag);\n  outputs.bounced(data, tag);\n} else {\n  const {timestamp} = data;\n  let buffer = state.buffer;\n  // Start of the new time window\n  let ts0 = timestamp - size;\n  let i0 = -1;\n  // Finding first non-overflowing sample\n  for (let i = 0, length = buffer.length; i < length; i++)\n    if (buffer[i].timestamp > ts0) {\n      i0 = i;\n      break;\n    }\n  buffer = \n    // All samples overflown\n    i0 === -1 ? [data]\n    // Some samples overflown\n    : i0 > 0 ? [...buffer.slice(i0), data]\n    // No samples overflown\n    : [...buffer, data];\n  state.buffer = buffer;\n  outputs.buffer(buffer, tag);\n}"
```

Comment:
I think size should be >= 0.
Also wouldn't it be simpler to simply use buffer.filter to filter out overflowing items?

Reply:
Still not clear to me why we use a for loop plus some conditional logic instead of Array.filter to remove overflown items from the buffer. I may not understand all the corner cases though...
If for some reason we stick to the method to find the first non-overflowing item, we still can use the Array.findIndex function like this:
`const i0 = buffer.findIndex((item) => item.timestamp > ts0)`
Nevertheless the current solution should also work fine, I just find it less readable then filter or findIndex.